### PR TITLE
Add docs

### PR DESCRIPTION
**Pull request by tomasz-sodzawiczny**

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+// Good, but avoid in most cases, only needed in low-level components.  
+wrapper.find('.class-name')
+```
+Bad:
+```js
+// Bad - selecting user-defined component by display name - use constructor.
+wrapper.find('SomeComponent')
+// Bad - selecting with string syntax - use filter and object syntax.
+wrapper.find('SomeComponent[someProps=value]')
+```
```
Comment (vergilius) [inline docs/index.md]: Lolo newline

File: docs/index.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
+# General
```
Comment (vergilius) [inline docs/index.md]: how about having table of contents?

File: docs/index.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+})
+// Bad - "should do" -> "does".
+it('shuold do something', () => {
+})
+// Bad - uppercase + '.'.
+it('Does something.', () => {
+})
+```
+
+### Code elements in tests descriptions
```
Comment (vergilius) [inline docs/index.md]: is syntax okay? gh stopped formatting `###` for me here

File: docs/index.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
+# General
+
+### Date/datetime handling
+Do not use `Date()` i `Date().now`, use helper functions instead.
```
Comment (kdzwinel) [inline docs/index.md]: `i` => `and`

Comment (kdzwinel) [reply docs/index.md]: What helper functions?

File: docs/index.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+  someFunction,
+  someOtherFunction,
+}
+```
+
+
+# React
+
+### `classNames`
+Use `classNames` helper always, when there is more than one class name. Use string for a single class name.
```
Comment (kdzwinel) [inline docs/index.md]: link to the classNames package / docs

Comment (kdzwinel) [reply docs/index.md]: hm, wait - if there are no conditional statements and list of classes is static why do we need `classNames`

Comment (kdzwinel) [reply docs/index.md]: @aju agrees

File: docs/index.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+```js
+  <MyComponent className={classNames('single-class')} />
+  
+  <MyComponent className="first-class second-class" />
+```
+
+### Using `ref`s
+Do not use `ref` to access a Component instance - we want to keep our components API as clean and closed as possible.
+
+It's ok to use `ref` for accessing a DOM elements:
```
Comment (kdzwinel) [inline docs/index.md]: ok -> OK, 'a DOM elements' -> 'DOM elements'

File: docs/index.md (hunk 3 above)
Comment (tomasz-sodzawiczny) [reply docs/index.md]: Works for me, makes a link and etc: https://github.com/brainly/frontend-tools-configs/blob/976571e732c6cabba7bbb9ed28dfff84997d0286/docs/index.md#code-elements-in-tests-descriptions

File: docs/index.md [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+    return (
+      <SomeChild onSomeEvent={this.handleSomeEvent} />
+    )
+  }
+```
+
+# Redux
+
+### Action format
+Use [Flux Standard Actions](https://github.com/acdlite/flux-standard-actios).
```
Comment (kdzwinel) [inline docs/index.md]: broken url

File: docs/index.md [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+const fetchSomeDataActionCreator = () => ...
+```
+
+### Async / sagas
+We use `thunk` middleware + `async`/`await` for actions with side-effects and sagas:
+```js
+const doSomeAsyncStuff = blah => {
+  return async dispatch => {
+    const result = await someAsyncStuff(blah)
+    dispatch(onAsyncStuffFinished(result))
```
Comment (kdzwinel) [inline docs/index.md]: missing semicolons - we are not savages!

File: docs/index.md [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+  it('does something', () => {
+  })
+})
+```
+Bad:
+```js
+// top level
+it('does something', () => {
+})
+```
```
Comment (kdzwinel) [inline docs/index.md]: these examples are almost the same as the ones above - not sure if we really need this section

File: docs/index.md [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+Keep test description simple and short, use correct English grammar.
+
+Keep in mind that nested describes + it should create a sentence, so:
+```js
+describe('A', () => {
+  describe('B', () => {
+    it('C', () => {
+     ...
+```
+results in a sentence: *"A B C"*
```
Comment (kdzwinel) [inline docs/index.md]: Can  you add some real example?

File: docs/index.md [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+// Bad - "should do" -> "does".
+it('shuold do something', () => {
+})
+// Bad - uppercase + '.'.
+it('Does something.', () => {
+})
+```
+
+### Code elements in tests descriptions
+Use names from code in descriptions, keep the same case: `SomeClass`, `someVariable`, `SOME_CONST`
```
Comment (kdzwinel) [inline docs/index.md]: missing `.` at the end

File: docs/index.md [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+    // Bad - this expect *always* passes.
+    // If the render crashes the test will fail on the previous line.
+    expect(wrapper).toHaveLength(1);
+  })
+})
+```
+
+### Enzyme `shallow` vs `mount`
+Avoid using `mount`, it usually means you're testing to much stuff in one test.
+#### Event handling
```
Comment (kdzwinel) [inline docs/index.md]: missing line break above

File: docs/index.md [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+`ShallowWrapper.simulate()` API does not mock events (unlike `mount`), but prefer to mock events yourself, than using `mount` for this reason:
```
Comment (kdzwinel) [inline docs/index.md]: hm, not sure what's wrong, but this sentence feels a bit off

File: docs/index.md [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+Good:
+```js
+wrapper = shallow(<MyComponent />)
+wrapper.simulate('click', {whatever: 'blah'})
+```
+Bad:
+```js
+wrapper = mount(<MyComponent />)
+wrapper.simulate('click')
+```
```
Comment (kdzwinel) [inline docs/index.md]: semicolons!

File: docs/index.md (hunk 9 above)
Comment (tomasz-sodzawiczny) [reply docs/index.md]: I'll actually merge these 2 points into one.

**Pull request merged**